New Seeking Behavior

rishiso opened the pull request:
## Description
Adding seek behavior to offense.

## Associated / Resolved Issue
Resolves [ClickUp](https://app.clickup.com/t/389zmk)

## Design Documents
[Link](link-to-design-doc)

## Steps to Test
make run-sim (forced offense to always search)

**Expected result:** Travels to open position

## Key Files to Review
 * offense.cpp
 * offense.hpp
 * seeking.cpp
 * seeking.hpp

## Review Checklist

- [x] **Docstrings**: All methods and classes should have the file appropriate docstrings which follow the guidelines in the ["Contributing" page](https://rj-rc-software.readthedocs.io/en/latest/contributing.html) of our docs.
- [x] **Remove extra print statements**: Any print statements used for debugging should be removed
- [x] **Tag reviewers**: Tag some people for review and ping them on Slack

## (Optional) Sub-issues (for drafts)
_Note: if you find yourself breaking this PR into many smaller features, it may make sense to break up the PR into logical units based on these features._
- [ ] Step 1
- [ ] Step 2

rishiso committed:
Accidentally broke old branch

kfu02 commented on soccer/src/soccer/strategy/agent/position/seeker.cpp:
```diff
@@ -0,0 +1,152 @@
+    rj_geometry::Point ball_pos = world_state->ball.position;
+    rj_geometry::Point min = current_point;
+    double min_val = max_los(ball_pos, current_point, world_state);
+    double curr_val;
```
```suggestion
    double curr_val{};
```
is safer to initialize variables, as C++ has unexpected behavior if this is ever uninitialized

kfu02 reviewed (changes requested):
![image](https://user-images.githubusercontent.com/30275369/232262793-6d0e4efe-e187-4425-9f93-fe62b81a7d6d.png)


This is what the seekers converged to after I let them stay put for a second, in general we don't want a clump of 3 robots in the same quadrant ever. I like the effort, but can you try going with a simpler model for now? Just something that spaces the Offense robots out roughly evenly around the opponent goal will do. For this upcoming comp we didn't have enough time to test and tune passing, so don't worry too much about all this movement to create open space.

If you want, you can leave this PR as a draft, but make a new one with the above recommendation implemented

rishiso committed:
Update variable initialization

Co-authored-by: Kevin Fu <30275369+kfu02@users.noreply.github.com>

rishiso committed:
Removing clumping

rishiso commented:
![Screenshot 2023-04-16 at 12 07 37 PM](https://user-images.githubusercontent.com/52169818/232326020-eaa50c45-24e3-4301-afc0-397223561e19.png)

I made some changes to reduce clumping. One of these changes is assigning individual bots to stay in a certain third based on position number. I also added more random noise and changed the model to incorporate max distance from friendly bots. The image above shows where the robots converge.

kfu02 reviewed (changes requested):
Looks better, nice! A few more things:

1. I had to modify your code to see the seeking behavior actually take place. Is this the case on your local branch? (If so, commit that change)
![image](https://user-images.githubusercontent.com/30275369/232330542-9ac6829c-7bbd-4069-8824-c039479cff36.png)

2. Docstrings should be placed in the .hpp files, this is just a convention with C++. It makes it easier for someone to simply read the header and understand what is happening. You have a few nice docstrings that are in the cpp files.

3. See the check warnings, in general we would prefer non-recursive calls if possible, so make that iterative instead (with some limit like 10 iterations, it won't improve much beyond that anyways). There's a couple minor ones about const/static, just listen to the warnings and look it up if you have time to learn.

Finally, I think the seek behavior is nice in sim, but we should refrain from confirming that it looks good without testing on real robots. So let's do that tonight.

rishiso committed:
Fixing docs and warnings

rishiso commented:
I have made the changes to the docstrings, moving them from the cpp to the header file. Also, I changed the recursive function to iterative. I am not seeing any more warnings for the seeker files, so I think that they have been resolved.

As for making the seeking behavior take place, I am forcing the current_state to always be searching similar to your code. I am not exactly sure how the offensive decision tree works, but it doesn't seem to be assigning the role of searching very often.

I won't be able to make it to tonight's meeting, but please let me know if the seeking works properly in real life.

sid-parikh commented:
@rishiso does this PR work still? can we clean it up and merge it?

rishiso commented:
@sid-parikh It is somewhat working. I was recreating the seeking behavior from the Python codebase in C++. However, there are a few issues since I had to recreate many Python functions from scratch like scipy.optimize. I can look and see if I can get it working fully.

sid-parikh commented:
@sanatd33 @shourikb You two have taken a card that involved finishing this PR. @rishiso if you think you can help them, or finish it quickly, you can do that as well. Let's aim to have this done in a couple meetings.

sid-parikh commented:
I have marked this as a draft for until it is done.
